src/control/server/mgmt_svc.go

```diff
@@ -800,87 +803,44 @@ func (svc *mgmtSvc) PrepShutdownRanks(ctx context.Context, req *mgmtpb.RanksReq)
-//
-// Iterate over instances issuing kill rank requests, wait until either all instances are
-// stopped or timeout occurs and then populate response results based on instance started state.
-// Return error if any instances are still running in order to enable retries at the sender.
-//
-// TODO: Enable "force" if number of retries fail, issuing a different signal/mechanism for
-//       terminating the process.
-func (svc *mgmtSvc) StopRanks(ctx context.Context, req *mgmtpb.RanksReq) (*mgmtpb.RanksResp, error) {
+// After attempting to stop instances through harness (when either all instances
+// are stopped or timeout has occurred, populate response results based on
+// instance started state.
+func (svc *mgmtSvc) StopRanks(parent context.Context, req *mgmtpb.RanksReq) (*mgmtpb.RanksResp, error) {
 	if req == nil {
 		return nil, errors.New("nil request")
 	}
 	svc.log.Debugf("MgmtSvc.StopRanks dispatch, req:%+v\n", *req)
 
 	resp := &mgmtpb.RanksResp{}
 
-	for _, i := range svc.harness.instances {
-		if !i.hasSuperblock() { // critical error
-			return nil, errors.Errorf("instance %d has no superblock", i.Index())
-		}
-
-		rank, ok := validateInstanceRank(svc.log, i, req.Ranks)
-		if !ok { // filtered out, no result expected
-			continue
-		}
-
-		if !i.IsStarted() { // skip as already stopped
-			svc.log.Debugf("rank %d already stopped", *rank)
-			continue
-		}
-
-		if err := i.Stop(req.Force); err != nil {
-			var msg string
-			if req.Force {
-				msg = " (forced)"
-			}
-			svc.log.Error(errors.Wrapf(err,
-				"rank %d stop%s", *rank, msg).Error())
-		}
+	var rankList []*ioserver.Rank
+	if err := convert.Types(req.Ranks, &rankList); err != nil {
+		return nil, errors.Wrap(err, "parsing request rank list")
 	}
 
-	stopped := make(chan struct{})
-	// select until instances stop or timeout occurs (at which point get results of each instance)
-	go func() {
-		for {
-			if len(svc.harness.StartedRanks()) > 0 {
-				time.Sleep(instanceUpdateDelay)
-				continue
-			}
-			close(stopped)
-			break
-		}
-	}()
-
-	select {
-	case <-stopped:
-	case <-time.After(svc.harness.rankReqTimeout * 2):
+	signal := syscall.SIGINT
+	if req.Force {
+		signal = syscall.SIGKILL
 	}
 
-	// either all instances stopped or timeout occurred
-	for _, i := range svc.harness.instances {
-		state := system.MemberStateStarted
-		rrErr := errors.Errorf("want %s, got %s", system.MemberStateStopped, state)
-
-		if !i.hasSuperblock() {
-			return nil, errors.Errorf("instance %d has no superblock", i.Index())
-		}
+	ctx, cancel := context.WithTimeout(parent, ioserverShutdownTimeout)
+	defer cancel()
 
-		if !i.IsStarted() {
-			state = system.MemberStateStopped
-			rrErr = nil
-		}
-		resp.Results = append(resp.Results,
-			NewRankResult(i.getSuperblock().Rank.Uint32(), "stop", state, rrErr))
+	results, err := svc.harness.StopInstances(ctx, signal, rankList...)
+	if err != nil && err != context.DeadlineExceeded {
```

Comment:
if err == context.DeadlineExceeded, do we need to do something else? For example, send a SIGKILL?

Reply:
at the moment this would be at the users discretion by reissuing the command with --force option, there might be occasions where we don't want to automatically SIGKILL if graceful shutdown fails